marlowe-playground-client/src/Halogen/CurrencyInput.js

```diff
@@ -0,0 +1,53 @@
+const { convertCompilerOptionsFromJson } = require("typescript");
+
+
+exports.setOldValues_ = function(input) {
+    input.old_value = input.value;
+    input.old_selectionStart = input.selectionStart;
+    input.old_selectionEnd = input.selectionEnd;
+}
+
+exports.checkChange_ = function(input) {
+    if (/^-?[0-9]*(\.[0-9]*)?$/.test(input.value)) {
+        exports.setOldValues_(input);
+    } else {
+        input.value = input.old_value;
+        input.selectionStart = input.old_selectionStart;
+        input.selectionEnd = input.old_selectionEnd;
+    }
+}
+
+exports.formatValueString_ = function(text, decimalPlaces) {
+    var positionSeparator = text.indexOf(".");
+    if (positionSeparator < 0) {
+        if (decimalPlaces > 0) {
+            text += '.';
+            text = text.padEnd(text.length + decimalPlaces, '0');
+        }
+    } else {
+        text = text.substring(0, positionSeparator + decimalPlaces + 1);
+        text = text.padEnd(positionSeparator + decimalPlaces + 1, '0');
+    }
+    if (text == '') {
+	text = '0';
+    } else if (text[0] == '.') {
+        text = '0' + text;
+    } else if ((text == '-') || ((text[0] == '-') && (text[1] == '.'))) {
+        text = '-0' + text.substring(1);
+    }
+    return text;
+}
```

Comment:
It seems that library uses `Number`, which would lose precision. I did consider yesterday too that, in retrospective, maybe I should have used purescript code for implementing this, but it is a small function anyway, and it is called from JS already, so it would require some extra boilerplate to go back and forwards, and it is so well tested 😇

Reply:
Maybe we could implement a `format` function that works with BigNumbers `https://github.com/purescript-contrib/purescript-formatters/blob/v5.0.1/src/Data/Formatter/Number.purs#L97-L97`. 

The back and forwards is if we want to keep the component in plain JS right? Which I would also change to be pure PS

Reply:
Yes, that would be good. In any case it is a bit of an independent issue to formatting in general, because having proper formatting in the input is much harder than having just a point and an optional minus sign at the beginning like I did (e.g: we could, and it would be good, to have thousand separators, for example). We can have better formatting outside the input, or you guy can try to improve the number input even further, but it doesn't seem easy, I studied the possibility for a bit